File: arc-agi3-kaggle-agent/arc_agent/world_model/mechanics_memory.py
"""[arc-agi3-kaggle-agent/world_model/mechanics_memory] BL.21561 -- memoria de mecanicas POR ACCION
y por EPISODIO, construida sobre `detectar_mecanica`. Puerto de
arc-agi-runner/src/worldModel/mechanicsMemory.ts.

POR ACCION acumula la firma de mecanica observada como distribucion Beta -- alpha son las veces que
la accion volvio a hacer LO MISMO, beta las veces que hizo otra cosa. Nunca un booleano: en
ARC-AGI-3 una accion de movimiento choca contra la pared cada tantos pasos y ahi no mueve nada, y
con verificacion de cero tolerancia esa unica observacion mataba la regla correcta.

POR EPISODIO implementa los dos detectores que no son de transicion:
4. MARCO/HUD: las celdas que no cambiaron NUNCA. Su complemento -- el bbox de lo que si cambio
   alguna vez -- es la ARENA: todo lo de afuera es decorado.
5. CONTADOR: un color cuya cantidad de celdas se mueve siempre en el mismo sentido es puntaje o
   vidas: senal densa de progreso.
"""
from __future__ import annotations

from dataclasses import dataclass
from typing import Final

from .grid import BoundingBox, Grid, VolatilityMask, is_volatile_cell
from .object_mechanics import Mecanica, Traslacion, detectar_mecanica
from .mechanics_signature import firma_de_mecanica

# Observaciones minimas de una accion antes de afirmar que su mecanica es conocida. 2 y no 1: una
# sola coincidencia no distingue una regla de una casualidad.
MIN_OBSERVACIONES_DE_MECANICA: Final[int] = 2

# Fraccion minima de observaciones que tienen que coincidir con la firma dominante. 0.6 deja pasar
# la regla de movimiento que falla contra la pared un tercio de las veces, y no la que acierta la
# mitad.
MIN_COBERTURA_DE_MECANICA: Final[float] = 0.6

# Cambios minimos de la cuenta de un color para llamarlo contador.
MIN_CAMBIOS_DE_CONTADOR: Final[int] = 3


@dataclass(frozen=True)
class HipotesisDeMecanica:
    action: str
    firma: str
    traslacion: Traslacion | None
    alpha: int
    beta: int
    observaciones: int
    cobertura: float


@dataclass(frozen=True)
class ContadorDeColor:
    color: int
    direccion: str
    cambios: int
    delta: int


class MechanicsMemory:
    def __init__(self) -> None:
        self._por_accion: dict[str, dict[str, object]] = {}
        self._cambio_alguna_vez: list[list[bool]] | None = None
        self._alto = 0
        self._ancho = 0
        self._observaciones_totales = 0
        self._conteo_anterior: dict[int, int] | None = None
        self._contadores: dict[int, dict[str, object]] = {}

    def observe(
        self, action: str, pre: Grid, post: Grid, mask: VolatilityMask | None = None
    ) -> Mecanica:
        """Registra el efecto de `action` y devuelve la mecanica detectada (logs y tests)."""
        mecanica = self._registrar_por_accion(action, pre, post, mask)
        self._registrar_celdas_cambiadas(pre, post, mask)
        self._registrar_contadores(post, mask)
        self._observaciones_totales += 1
        return mecanica

    def observe_evidencia_adicional(
        self, action: str, pre: Grid, post: Grid, mask: VolatilityMask | None = None
    ) -> Mecanica:
        """BL.22236 -- variante de `observe()` para evidencia INTERMEDIA: una transicion entre dos
        capas de animacion del MISMO frame (`state_signature.extraer_grid_multicapa`), no el
        pre/post asentado de la accion.

        Actualiza SOLO el detector 3 (hipotesis de mecanica por-accion, via
        `_registrar_por_accion`) y NO los detectores 4/5 (arena / contador), que son POR EPISODIO
        y describen el tablero ASENTADO: una celda que aparece y desaparece durante una animacion
        de "pouring" no es parte de la arena jugable ni de un contador de progreso monotono -- es
        evidencia de la MECANICA de la accion, que es justo lo que 13/25 juegos publicos esconden
        en capas intermedias. La hipotesis por-accion ya tolera evidencia ruidosa por diseno
        (Beta + `MIN_COBERTURA_DE_MECANICA=0.6`), asi que sumar observaciones intermedias no puede
        voltear una regla bien establecida por una minoria de capas transitorias."""
        return self._registrar_por_accion(action, pre, post, mask)

    def _registrar_por_accion(
        self, action: str, pre: Grid, post: Grid, mask: VolatilityMask | None
    ) -> Mecanica:
        mecanica = detectar_mecanica(pre, post, mask)
        firma = firma_de_mecanica(mecanica)

        registro = self._por_accion.setdefault(
            action, {"conteo": {}, "traslaciones": {}, "observaciones": 0}
        )
        registro["observaciones"] = int(registro["observaciones"]) + 1
        conteo: dict[str, int] = registro["conteo"]  # type: ignore[assignment]
        conteo[firma] = conteo.get(firma, 0) + 1
        if mecanica.traslacion_principal is not None:
            traslaciones: dict[str, Traslacion] = registro["traslaciones"]  # type: ignore[assignment]
            traslaciones[firma] = mecanica.traslacion_principal
        return mecanica

    def get_hypothesis(self, action: str) -> HipotesisDeMecanica | None:
        """Firma mas observada de `action` con su Beta. None si la accion nunca se observo."""
        registro = self._por_accion.get(action)
        if registro is None:
            return None
        conteo: dict[str, int] = registro["conteo"]  # type: ignore[assignment]
        # Desempate por firma (orden lexicografico) y no por orden de insercion: dos firmas
        # empatadas tienen que resolver igual aca y en el motor TypeScript.
        firma = sorted(conteo.items(), key=lambda kv: (-kv[1], kv[0]))[0][0]
        aciertos = conteo[firma]
        observaciones = int(registro["observaciones"])
        traslaciones: dict[str, Traslacion] = registro["traslaciones"]  # type: ignore[assignment]
        return HipotesisDeMecanica(
            action=action,
            firma=firma,
            traslacion=traslaciones.get(firma),
            alpha=1 + aciertos,
            beta=1 + observaciones - aciertos,
            observaciones=observaciones,
            cobertura=aciertos / observaciones,
        )

    def get_direction(self, action: str) -> tuple[int, int] | None:
        """Direccion (dy,dx) CONFIRMADA de `action`, o None si no mueve nada / falta evidencia. Es
        el mapeo ACTION1..5 -> direccion que el DSL global nunca pudo dar sobre dato real."""
        h = self.get_hypothesis(action)
        if h is None or h.traslacion is None:
            return None
        if not h.firma.startswith("traslacion:"):
            return None
        if h.observaciones < MIN_OBSERVACIONES_DE_MECANICA:
            return None
        if h.cobertura < MIN_COBERTURA_DE_MECANICA:
            return None
        return (h.traslacion.dy, h.traslacion.dx)

    def get_movement_actions(self) -> list[str]:
        return [a for a in self._por_accion if self.get_direction(a) is not None]

    def is_inert_action(self, action: str) -> bool:
        """Accion cuya mecanica dominante es `sinCambio` con evidencia suficiente -- no-op
        observacional, sin pasar por la sintesis DSL."""
        h = self.get_hypothesis(action)
        if h is None:
            return False
        return (
            h.firma == "sinCambio"
            and h.observaciones >= MIN_OBSERVACIONES_DE_MECANICA
            and h.cobertura >= MIN_COBERTURA_DE_MECANICA
        )

    def get_active_bounding_box(self) -> BoundingBox | None:
        """DETECTOR 4 -- caja de lo que cambio alguna vez: la ARENA. Todo lo de afuera es marco
        estatico. None mientras no se observo ningun cambio."""
        if self._cambio_alguna_vez is None:
            return None
        ys: list[int] = []
        xs: list[int] = []
        for y in range(self._alto):
            for x in range(self._ancho):
                if self._cambio_alguna_vez[y][x]:
                    ys.append(y)
                    xs.append(x)
        if not ys:
            return None
        return BoundingBox(min_y=min(ys), max_y=max(ys), min_x=min(xs), max_x=max(xs))

    def get_static_cell_count(self) -> int:
        """DETECTOR 4 -- celdas que no cambiaron NUNCA en todo lo observado."""
        if self._cambio_alguna_vez is None:
            return 0
        return sum(1 for fila in self._cambio_alguna_vez for c in fila if not c)

    def is_static_cell(self, y: int, x: int) -> bool:
        if self._cambio_alguna_vez is None:
            return True
        if y < 0 or x < 0 or y >= self._alto or x >= self._ancho:
            return True
        return not self._cambio_alguna_vez[y][x]

    def get_counters(self) -> list[ContadorDeColor]:
        """DETECTOR 5 -- colores cuya cantidad de celdas se movio SIEMPRE en el mismo sentido."""
        salida: list[ContadorDeColor] = []
        for color in sorted(self._contadores):
            c = self._contadores[color]
            if c["roto"] or c["direccion"] is None:
                continue
            if int(c["cambios"]) < MIN_CAMBIOS_DE_CONTADOR:
                continue
            salida.append(
                ContadorDeColor(
                    color=color,
                    direccion=str(c["direccion"]),
                    cambios=int(c["cambios"]),
                    delta=int(c["delta"]),
                )
            )
        salida.sort(key=lambda c: (-c.cambios, c.color))
        return salida

    def get_observation_count(self) -> int:
        return self._observaciones_totales

    def _registrar_celdas_cambiadas(
        self, pre: Grid, post: Grid, mask: VolatilityMask | None
    ) -> None:
        alto = len(pre)
        ancho = len(pre[0]) if pre else 0
        if alto == 0 or ancho == 0:
            return
        # Un cambio de forma del frame reinicia el mapa: las coordenadas viejas ya no describen el
        # mismo tablero, y mezclarlas produciria una arena inventada.
        if self._cambio_alguna_vez is None or self._alto != alto or self._ancho != ancho:
            self._cambio_alguna_vez = [[False] * ancho for _ in range(alto)]
            self._alto = alto
            self._ancho = ancho
        for y in range(alto):
            fila = pre[y]
            fila_post = post[y] if y < len(post) else []
            for x in range(min(len(fila), ancho)):
                valor_post = fila_post[x] if x < len(fila_post) else None
                if fila[x] != valor_post and not is_volatile_cell(mask, y, x):
                    self._cambio_alguna_vez[y][x] = True

    def _registrar_contadores(self, post: Grid, mask: VolatilityMask | None) -> None:
        conteo: dict[int, int] = {}
        for y in range(len(post)):
            fila = post[y]
            for x in range(len(fila)):
                if is_volatile_cell(mask, y, x):
                    continue
                conteo[fila[x]] = conteo.get(fila[x], 0) + 1
        anterior = self._conteo_anterior
        self._conteo_anterior = conteo
        if anterior is None:
            return
        for color in set(anterior) | set(conteo):
            antes = anterior.get(color, 0)
            ahora = conteo.get(color, 0)
            if antes == ahora:
                continue
            direccion = "sube" if ahora > antes else "baja"
            estado = self._contadores.setdefault(
                color, {"direccion": None, "cambios": 0, "delta": 0, "roto": False}
            )
            if estado["direccion"] is not None and estado["direccion"] != direccion:
                estado["roto"] = True
            if estado["direccion"] is None:
                estado["direccion"] = direccion
            estado["cambios"] = int(estado["cambios"]) + 1
            estado["delta"] = int(estado["delta"]) + (ahora - antes)

File: arc-agi-runner/src/worldModel/mechanicsMemory.ts
/* [arc-agi-runner/worldModel/mechanicsMemory] BL.21561 -- memoria de mecanicas POR ACCION y por
   EPISODIO, construida sobre `detectarMecanica` (objectMechanics.ts).

   POR ACCION acumula la firma de mecanica observada como distribucion Beta -- alpha son las veces
   que la accion volvio a hacer LO MISMO, beta las veces que hizo otra cosa. Nunca un booleano: en
   ARC-AGI-3 una accion de movimiento choca contra la pared cada tantos pasos y ahi no mueve nada;
   con verificacion de cero tolerancia esa unica observacion mataba la regla correcta (que es
   exactamente lo que rompia `verifyProgram`, ver synthesis.ts).

   POR EPISODIO implementa los dos detectores que no son de transicion:
   4. MARCO/HUD: las celdas que no cambiaron NUNCA. Su complemento -- el bbox de lo que si cambio
      alguna vez -- es la ARENA: todo lo de afuera es decorado y no deberia entrar en ninguna
      decision.
   5. CONTADOR: un color cuya cantidad de celdas se mueve siempre en el mismo sentido es puntaje o
      vidas -- senal densa de progreso, a diferencia del score del juego que solo cambia al ganar
      un nivel. */

import { isVolatileCell, type BoundingBox, type Grid, type VolatilityMask } from './grid';
import { firmaDeMecanica } from './mechanicsSignature';
import { detectarMecanica, type Mecanica, type Traslacion } from './objectMechanics';
// BL.21741 (correccion): la capa de VOCABULARIO vive en su propio modulo -- detectar y nombrar son
// dos responsabilidades, y objectMechanics.ts cruzaba el limite de 500 lineas del repo.

/** Observaciones minimas de una accion antes de afirmar que su mecanica es conocida. 2 y no 1:
 *  una sola coincidencia no distingue una regla de una casualidad, y en ARC-AGI-3 hay acciones
 *  (el click) que hacen cosas distintas segun donde caen. */
export const MIN_OBSERVACIONES_DE_MECANICA = 2;

/** Fraccion minima de observaciones que tienen que coincidir con la firma dominante para
 *  considerar que la accion TIENE una mecanica. 0.6 deja pasar la regla de movimiento que falla
 *  contra la pared un tercio de las veces, y no la que acierta la mitad. */
export const MIN_COBERTURA_DE_MECANICA = 0.6;

/** Cambios minimos de la cuenta de un color para llamarlo contador. Con menos, cualquier objeto
 *  que aparece y se queda quieto pasaria por marcador. */
export const MIN_CAMBIOS_DE_CONTADOR = 3;

export interface HipotesisDeMecanica {
  action: string;
  /** Firma dominante -- `sinCambio`, `traslacion:dy,dx`, `recoloreo:a>b`, ... */
  firma: string;
  /** Traslacion asociada a la firma dominante (`null` si la firma no es una traslacion). */
  traslacion: Traslacion | null;
  alpha: number;
  beta: number;
  observaciones: number;
  /** alpha-1 sobre observaciones: fraccion de las veces que la accion hizo lo mismo. */
  cobertura: number;
}

export interface ContadorDeColor {
  color: number;
  direccion: 'sube' | 'baja';
  cambios: number;
  delta: number;
}

interface RegistroDeAccion {
  conteoPorFirma: Map<string, number>;
  ultimaTraslacionPorFirma: Map<string, Traslacion>;
  observaciones: number;
}

export class MechanicsMemory {
  private readonly porAccion = new Map<string, RegistroDeAccion>();
  /** `true` en las celdas que cambiaron alguna vez -- el complemento es el marco estatico. */
  private cambioAlgunaVez: Uint8Array | null = null;
  private alto = 0;
  private ancho = 0;
  private observacionesTotales = 0;
  /** Cuenta de celdas por color en el ultimo frame observado. */
  private conteoAnterior: Map<number, number> | null = null;
  private readonly contadores = new Map<
    number,
    { direccion: 'sube' | 'baja' | null; cambios: number; delta: number; roto: boolean }
  >();

  /** Registra el efecto de `action` y devuelve la mecanica detectada (para logs y tests). */
  observe(action: string, pre: Grid, post: Grid, mask: VolatilityMask | null = null): Mecanica {
    const mecanica = detectarMecanica(pre, post, mask);
    const firma = firmaDeMecanica(mecanica);

    const registro = this.porAccion.get(action) ?? {
      conteoPorFirma: new Map<string, number>(),
      ultimaTraslacionPorFirma: new Map<string, Traslacion>(),
      observaciones: 0,
    };
    registro.observaciones += 1;
    registro.conteoPorFirma.set(firma, (registro.conteoPorFirma.get(firma) ?? 0) + 1);
    if (mecanica.traslacionPrincipal !== null) {
      registro.ultimaTraslacionPorFirma.set(firma, mecanica.traslacionPrincipal);
    }
    this.porAccion.set(action, registro);

    this.registrarCeldasCambiadas(pre, post, mask);
    this.registrarContadores(post, mask);
    this.observacionesTotales += 1;
    return mecanica;
  }

  /** Hipotesis vigente de `action`: la firma mas observada, con su Beta. `undefined` si la accion
   *  nunca se observo. */
  getHypothesis(action: string): HipotesisDeMecanica | undefined {
    const registro = this.porAccion.get(action);
    if (registro === undefined) return undefined;
    /* Desempate por firma (orden lexicografico) y no por orden de insercion: dos firmas empatadas
       tienen que resolver igual en el motor TS y en el puerto Python, que no comparten el orden
       de iteracion de sus mapas. */
    const ordenadas = [...registro.conteoPorFirma.entries()].sort(
      (a, b) => b[1] - a[1] || (a[0] < b[0] ? -1 : a[0] > b[0] ? 1 : 0),
    );
    const [firma, aciertos] = ordenadas[0];
    const fallos = registro.observaciones - aciertos;
    return {
      action,
      firma,
      traslacion: registro.ultimaTraslacionPorFirma.get(firma) ?? null,
      alpha: 1 + aciertos,
      beta: 1 + fallos,
      observaciones: registro.observaciones,
      cobertura: aciertos / registro.observaciones,
    };
  }

  /** Direccion de movimiento CONFIRMADA de `action` -- `null` si la accion no mueve nada, si aun
   *  no hay evidencia suficiente o si la mecanica dominante no es una traslacion. Es el mapeo
   *  ACTION1..5 -> direccion que el DSL global nunca pudo dar. */
  getDirection(action: string): { dy: number; dx: number } | null {
    const h = this.getHypothesis(action);
    if (h === undefined || h.traslacion === null) return null;
    if (!h.firma.startsWith('traslacion:')) return null;
    if (h.observaciones < MIN_OBSERVACIONES_DE_MECANICA) return null;
    if (h.cobertura < MIN_COBERTURA_DE_MECANICA) return null;
    return { dy: h.traslacion.dy, dx: h.traslacion.dx };
  }

  /** Acciones con direccion de movimiento confirmada, en orden de registro. */
  getMovementActions(): string[] {
    return [...this.porAccion.keys()].filter((a) => this.getDirection(a) !== null);
  }

  /** Acciones con mecanica dominante `sinCambio` y evidencia suficiente -- no-op observacional,
   *  sin pasar por la sintesis DSL. */
  isInertAction(action: string): boolean {
    const h = this.getHypothesis(action);
    if (h === undefined) return false;
    return (
      h.firma === 'sinCambio' &&
      h.observaciones >= MIN_OBSERVACIONES_DE_MECANICA &&
      h.cobertura >= MIN_COBERTURA_DE_MECANICA
    );
  }

  /** DETECTOR 4 -- caja de lo que cambio alguna vez: la ARENA. Todo lo de afuera es marco/HUD
   *  estatico. `null` mientras no se observo ningun cambio. */
  getActiveBoundingBox(): BoundingBox | null {
    if (this.cambioAlgunaVez === null) return null;
    let minY = Infinity;
    let minX = Infinity;
    let maxY = -Infinity;
    let maxX = -Infinity;
    for (let y = 0; y < this.alto; y++) {
      for (let x = 0; x < this.ancho; x++) {
        if (this.cambioAlgunaVez[y * this.ancho + x] === 1) {
          if (y < minY) minY = y;
          if (y > maxY) maxY = y;
          if (x < minX) minX = x;
          if (x > maxX) maxX = x;
        }
      }
    }
    if (maxY < minY) return null;
    return { minY, minX, maxY, maxX };
  }

  /** DETECTOR 4 -- celdas que no cambiaron NUNCA en todo lo observado. */
  getStaticCellCount(): number {
    if (this.cambioAlgunaVez === null) return 0;
    let estaticas = 0;
    for (const marca of this.cambioAlgunaVez) if (marca === 0) estaticas++;
    return estaticas;
  }

  isStaticCell(y: number, x: number): boolean {
    if (this.cambioAlgunaVez === null) return true;
    if (y < 0 || x < 0 || y >= this.alto || x >= this.ancho) return true;
    return this.cambioAlgunaVez[y * this.ancho + x] === 0;
  }

  /** DETECTOR 5 -- colores cuya cantidad de celdas se movio SIEMPRE en el mismo sentido:
   *  puntaje/vidas. Ordenado por cantidad de cambios descendente. */
  getCounters(): ContadorDeColor[] {
    const salida: ContadorDeColor[] = [];
    for (const [color, c] of [...this.contadores.entries()].sort((a, b) => a[0] - b[0])) {
      if (c.roto || c.direccion === null) continue;
      if (c.cambios < MIN_CAMBIOS_DE_CONTADOR) continue;
      salida.push({ color, direccion: c.direccion, cambios: c.cambios, delta: c.delta });
    }
    return salida.sort((a, b) => b.cambios - a.cambios || a.color - b.color);
  }

  getObservationCount(): number {
    return this.observacionesTotales;
  }

  private registrarCeldasCambiadas(pre: Grid, post: Grid, mask: VolatilityMask | null): void {
    const alto = pre.length;
    const ancho = pre[0]?.length ?? 0;
    if (alto === 0 || ancho === 0) return;
    /* Un cambio de forma del frame reinicia el mapa: las coordenadas viejas ya no describen el
       mismo tablero, y mezclarlas produciria una arena inventada. */
    if (this.cambioAlgunaVez === null || this.alto !== alto || this.ancho !== ancho) {
      this.cambioAlgunaVez = new Uint8Array(alto * ancho);
      this.alto = alto;
      this.ancho = ancho;
    }
    for (let y = 0; y < alto; y++) {
      const fila = pre[y];
      const filaPost = post[y] ?? [];
      for (let x = 0; x < fila.length && x < ancho; x++) {
        if (fila[x] !== filaPost[x] && !isVolatileCell(mask, y, x)) {
          this.cambioAlgunaVez[y * ancho + x] = 1; // @proto-safe: Uint8Array indexado por entero acotado, sin claves de string
        }
      }
    }
  }

  private registrarContadores(post: Grid, mask: VolatilityMask | null): void {
    const conteo = new Map<number, number>();
    for (let y = 0; y < post.length; y++) {
      const fila = post[y];
      for (let x = 0; x < fila.length; x++) {
        if (isVolatileCell(mask, y, x)) continue;
        conteo.set(fila[x], (conteo.get(fila[x]) ?? 0) + 1);
      }
    }
    const anterior = this.conteoAnterior;
    this.conteoAnterior = conteo;
    if (anterior === null) return;

    const colores = new Set([...anterior.keys(), ...conteo.keys()]);
    for (const color of colores) {
      const antes = anterior.get(color) ?? 0;
      const ahora = conteo.get(color) ?? 0;
      if (antes === ahora) continue;
      const direccion: 'sube' | 'baja' = ahora > antes ? 'sube' : 'baja';
      const estado = this.contadores.get(color) ?? {
        direccion: null,
        cambios: 0,
        delta: 0,
        roto: false,
      };
      if (estado.direccion !== null && estado.direccion !== direccion) estado.roto = true;
      estado.direccion = estado.direccion ?? direccion;
      estado.cambios += 1;
      estado.delta += ahora - antes;
      this.contadores.set(color, estado);
    }
  }
}
